src/modules/uORB/uORB_tests/uORBTest_UnitTest.cpp

```diff
@@ -582,6 +588,148 @@ int uORBTest::UnitTest::test_multi_reversed()
 	return test_note("PASS multi-topic reversed");
 }
 
+int uORBTest::UnitTest::test_wrap_around()
+{
+	test_note("Testing orb wrap-around");
+
+	orb_test_medium_s t{};
+	orb_test_medium_s u{};
+	orb_advert_t ptopic{nullptr};
+	bool updated{false};
+
+	int sfd = orb_subscribe(ORB_ID(orb_test_medium_wrap_around));
+
+	if (sfd < 0) {
+		return test_fail("subscribe failed: %d", errno);
+	}
+
+	const int queue_size = 16;
+	t.val = 0;
+	ptopic = orb_advertise_queue(ORB_ID(orb_test_medium_wrap_around), &t, queue_size);
+
+	if (ptopic == nullptr) {
+		return test_fail("advertise failed: %d", errno);
+	}
+
+	// Set generation to the location where wrap-around is about to be
+	{
+		auto node = uORB::Manager::get_instance()->get_device_master()->getDeviceNode(ORB_ID(orb_test_medium_wrap_around), 0);
+
+		if (node == nullptr) {
+			return test_fail("get device node failed.");
+		}
+
+		// Set to (0 - queue_size / 2) => (UINT_MAX - queue_size / 2 + 1)
+		set_generation(*node, unsigned(-(queue_size / 2)));
+
+		if (node->updates_available(0) != unsigned(-(queue_size / 2))) {
+			return test_fail("set generation failed.");
+		}
+
+		// Refresh the subscriber's generation
+		for (int i = 0; i < queue_size; i++) {
+			orb_check(sfd, &updated);
+
+			if (!updated) {
+				break;
+			}
+
+			orb_copy(ORB_ID(orb_test_medium_wrap_around), sfd, &u);
+		}
+	}
+
+	orb_publish(ORB_ID(orb_test_medium_wrap_around), ptopic, &t);
+
+	orb_check(sfd, &updated);
+
+	if (!updated) {
+		return test_fail("update flag not set");
+	}
+
+	if (PX4_OK != orb_copy(ORB_ID(orb_test_medium_wrap_around), sfd, &u)) {
+		return test_fail("copy(1) failed: %d", errno);
+	}
+
+	if (u.val != t.val) {
+		return test_fail("copy(1) mismatch: %d expected %d", u.val, t.val);
+	}
+
+	orb_check(sfd, &updated);
+
+	if (updated) {
+		return test_fail("spurious updated flag");
+	}
+
+#define CHECK_UPDATED(element) \
+	orb_check(sfd, &updated); \
+	if (!updated) { \
+		return test_fail("update flag not set, element %i", element); \
+	}
+#define CHECK_NOT_UPDATED(element) \
+	orb_check(sfd, &updated); \
+	if (updated) { \
+		return test_fail("update flag set, element %i", element); \
+	}
+#define CHECK_COPY(i_got, i_correct) \
+	orb_copy(ORB_ID(orb_test_medium_wrap_around), sfd, &u); \
+	if (i_got != i_correct) { \
+		return test_fail("got wrong element from the queue (got %i, should be %i)", i_got, i_correct); \
+	}
+
+	//no messages in the queue anymore
+
+	test_note("  Testing to write some elements...");
+
+	for (int i = 0; i < queue_size - 2; ++i) {
+		t.val = i;
+		orb_publish(ORB_ID(orb_test_medium_wrap_around), ptopic, &t);
+	}
+
+	for (int i = 0; i < queue_size - 2; ++i) {
+		CHECK_UPDATED(i);
+		CHECK_COPY(u.val, i);
+	}
+
+	CHECK_NOT_UPDATED(queue_size);
+
+	test_note("  Testing overflow...");
```

Comment:
Here you would need to reset the generation

Reply:
Now it's done.